crates/provider/src/layers/cache.rs

```diff
@@ -0,0 +1,150 @@
+use crate::{Provider, ProviderLayer, RootProvider};
+use alloy_json_rpc::{Request, RpcParam};
+use alloy_network::Ethereum;
+use alloy_primitives::B256;
+use alloy_rpc_types_eth::{Block, BlockNumberOrTag};
+use alloy_transport::{Transport, TransportErrorKind, TransportResult};
+use lru::LruCache;
+use std::{marker::PhantomData, num::NonZeroUsize, path::PathBuf};
+use tokio::sync::Mutex;
+
+// TODO: Populate load cache from file on initialization.
+// TODO: Add method to dump cache to file.
+/// A provider layer that caches RPC responses and serves them on subsequent requests.
+#[derive(Debug, Clone)]
+pub struct CacheLayer {
+    path: PathBuf,
+    max_items: usize,
```

Comment:
we can put these types into a config struct, so we can easily extend this, perhaps this should even store the actual cache

Reply:
+1 on moving it to a config struct. But don't think it should store the actual cache since that would requires us to:

```rust
struct CacheLayer {
cache: RwLock<Lru>,
}

struct CacheProvider {
layer: CacheLayer
}
```

This is an anti-pattern as the user initializes the layer like so `.layer(CacheLayer::new(path, max_items)` while initializing the `Provider` using `ProviderBuilder`. And in turn, the `CacheProvider` is initialized.